### PR TITLE
feat: add support for global allocatable array

### LINKED ISSUE #3993: Allocatables in modules aren't defined
```fortran

module a
	implicit none
	integer ,allocatable , dimension(:) :: f 
end module a


program mm
	use a
	implicit none
	allocate(f(4))
	f = [1,2,3,4]
	print *,f
end program mm
```

```shell
code generation error: Variable type not supported 15
 --> first.f90:3:41
  |
3 |  integer ,allocatable , dimension(:) :: f 
  |                                         ^ 

```

### PR DESCRIPTION
**Pull request by assem2002**
closes #3993 

The main issue this PR handles is how we would initialize the array pointer.
I chose to initialize it using `fill_array_details_` inside the main function as stays there in the stack memory through the whole life of the program, so it won't make a problem initializing the `array-type (struct)` inside it.

Another approach is to create global array type (struct) in the global scope, but that would require creating a whole `fill_array_details` to write the API instructions.

### CODE REVIEW

**Comment by assem2002** [comment]
The SNAP Package still throw the same error. 
I believe it's due to the external symbol assertion failing as the pointer to array is still set to null at the time of processing the mod files (as I think).

I'll look into it.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -2742,6 +2749,14 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
             llvm::Constant *ptr = module->getOrInsertGlobal(x.m_name,
                 x_ptr);
+            if (x.m_type->type == ASR::ttypeType::Allocatable){
+                llvm::Type* type_ = llvm_utils->get_type_from_ttype_t_util(
+                            ASRUtils::type_get_past_allocatable(x.m_type), module.get(), x.m_abi);
+                allocatable_array_details.push_back({ptr,
+                        type_,
+                        x.m_type,
+                        (down_cast<ASR::Array_t>((down_cast<ASR::Allocatable_t>(x.m_type))->m_type))->n_dims});
+            }
```
Comment (Pranavchiku) [inline src/libasr/codegen/asr_to_llvm.cpp:2759]: nice, approach, this is along right directions.

File: integration_tests/global_allocatable_02.f90 [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    logical :: ret
+    allocate(f(4,4))
+    f(1,:) = [1, 2, 3, 4]
+    print *,f
+    call r
+    print *,f
+    ret = all (f(2,:) == [10,10,10,10,10,10])
+    if(ret .eqv. .false.) error stop
+
+ end program mm
```
Comment (Pranavchiku) [inline integration_tests/global_allocatable_02.f90:30]: good test, I had a suspicion that the approach won't work if we used `f` in different routines. This test proves that it works.

**Review by Pranavchiku — APPROVED** [review]
Brilliant work! Thanks for this PR!

**Comment by Pranavchiku** [comment]
This PR uses approach that I mentioned here https://github.com/lfortran/lfortran/pull/3997#issuecomment-2138769107, so I think this is along right direction. Changes look good to me.

Rebase this PR, and then it will be good to merge.

**Comment by assem2002** [comment]
@Pranavchiku  Thank you. I think it's ready to be merged now.

**Comment by Pranavchiku** [comment]
Please do not merge PRs with merge commit in history, also lookup commit history before merging. We do not update PRs through merge commit, we rebase it.

**Comment by assem2002** [comment]
Okay. Sorry for that. I'll revert it now.

**Comment by Pranavchiku** [comment]
Well, let's wait, I'll let Ubaid, or Ondrej make a decision.

**Comment by certik** [comment]
Don't revert, that would make a mess too. Next time either squash, or rebase the history to make it clean.

**Comment by certik** [comment]
Is the idea here to create the array descriptor in the main program, so that it can be allocated later on?

**Comment by assem2002** [comment]
> Is the idea here to create the array descriptor in the main program, so that it can be allocated later on?

The declaration of allocatable result in this code
```Fortran
```

**Pull request merged**